hydradx/model/amm/arbitrage_agent_general.py

```diff
@@ -0,0 +1,549 @@
+    and virtual agents to keep track of the results.
+    """
+    buffer = swap_config['buffer']
+    exchange_names = list(swap_config['exchanges'].keys())
+    tkn_pairs = swap_config['exchanges']
+    slippage_tolerance = {ex_name: buffer / 2 for ex_name in exchange_names}
+
+    # make sure the assets are specified in the correct order
+    for ex_name in swap_config['exchanges']:
+        if isinstance(exchanges[ex_name], CentralizedMarket):
+            if tkn_pairs[ex_name] not in exchanges[ex_name].order_book:
+                raise ValueError(f'Asset pair {tkn_pairs[ex_name]} not found in order book. Are they ordered wrong?')
+
+    # all denominated in tkn_pair[1]
+    buy_price = {
+        **{ex_name: exchanges[ex_name].buy_spot(
+            tkn_buy=tkn_pairs[ex_name][0], tkn_sell=tkn_pairs[ex_name][1]
+        ) for ex_name in exchange_names},
+        'feeless': {ex_name: exchanges[ex_name].buy_spot(
+            tkn_buy=tkn_pairs[ex_name][0], tkn_sell=tkn_pairs[ex_name][1], fee=0
+        ) for ex_name in exchange_names}
+    }
+    sell_price = {
+        **{ex_name: exchanges[ex_name].sell_spot(
+            tkn_buy=tkn_pairs[ex_name][1], tkn_sell=tkn_pairs[ex_name][0]
+        ) for ex_name in exchange_names},
+        'feeless': {ex_name: exchanges[ex_name].sell_spot(
+            tkn_buy=tkn_pairs[ex_name][1], tkn_sell=tkn_pairs[ex_name][0], fee=0
+        ) for ex_name in exchange_names}
+    }
+    swap = {}
+
+    buy_ex, sell_ex = None, None
+    for ex_name, other_ex in (exchange_names, reversed(exchange_names)):
+        if buy_price[ex_name] and sell_price[other_ex] and buy_price[ex_name] < sell_price[other_ex]:
+            buy_ex = ex_name
+            sell_ex = other_ex
+            break
+
+    if buy_ex and sell_ex:
+        amt = calculate_arb_amount(
+            buy_ex=exchanges[buy_ex], sell_ex=exchanges[sell_ex],
+            buy_ex_tkn_pair=tkn_pairs[buy_ex], sell_ex_tkn_pair=tkn_pairs[sell_ex],
+            buffer=buffer, min_amt=1e-6,
+            buy_ex_max_sell=max_liquidity[buy_ex][tkn_pairs[buy_ex][0]],
+            sell_ex_max_sell=max_liquidity[sell_ex][tkn_pairs[sell_ex][0]],
+            precision=precision,
+            max_iters=max_iters
+        )
+
+        if amt != 0:
+            init_amt = {
+                buy_ex: agents[buy_ex].holdings[tkn_pairs[buy_ex][1]],
+                sell_ex: agents[sell_ex].holdings[tkn_pairs[sell_ex][1]]
+            }
+            exchanges[buy_ex].swap(
+                agent=agents[buy_ex],
+                tkn_buy=tkn_pairs[buy_ex][0],
+                tkn_sell=tkn_pairs[buy_ex][1],
+                buy_quantity=amt
+            )
+            exchanges[sell_ex].swap(
+                agent=agents[sell_ex],
+                tkn_buy=tkn_pairs[sell_ex][1],
+                tkn_sell=tkn_pairs[sell_ex][0],
+                sell_quantity=amt
+            )
+            amt_in = {
```

Comment:
I changed it so the default holdings are 1,000,000,000. It's maybe not ideal because it could increase rounding errors, but it is simple and it works. Let me know if you have a better idea.

Reply:
I think the best solution is this: There's no reason the agent that you call swap with on lines 159 and 165 have to be the same agent in the input of process_next_swap. I would instantiate new agents right before these swap() function calls, with holdings just above what they need to be to execute the swaps. This will allow you to calculate amt_in and amt_out with maximal precision.

Note that you may be able to remove the agents input from the function altogether, I'm not sure. You may want to leave it to check that the buy_ex and sell_ex agents do have enough of the required assets to execute the trades, I'm not sure if this is the where that should be done or not. But there's no reason for the buy_ex and sell_ex agents themselves to be what you pass into the swap functions on lines 159 and 165.

Generally speaking, this fixes the issue of precision depending on the initial agents holdings in a way that it doesn't need to.

Reply:
This works. Instantiating new agents for every swap will slow it down some, but it may not be noticeable.